ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ExplicitDisclosureIT.scala

```diff
@@ -0,0 +1,421 @@
+        .exerciseFetchDelegated(
+          testContext.disclosedContract
+            .update(_.metadata.createdAt := com.google.protobuf.timestamp.Timestamp.of(1, 0))
+        )
+        .mustFail("using a mismatching ledger time")
+
+      // Exercise a choice using invalid explicit disclosure (bad payload)
+      errorBadPayload <- testContext
+        .exerciseFetchDelegated(
+          testContext.disclosedContract
+            .update(_.arguments := Delegated(delegate, testContext.contractKey).arguments)
+        )
+        .mustFail("using an invalid disclosed contract payload")
+    } yield {
+      assertGrpcError(
+        errorMismatchingContractId,
+        LedgerApiErrors.ConsistencyErrors.DisclosedContractInvalid,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+      assertGrpcError(
+        errorBadKey,
+        LedgerApiErrors.ConsistencyErrors.DisclosedContractInvalid,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+      assertGrpcError(
+        errorBadLet,
+        LedgerApiErrors.ConsistencyErrors.DisclosedContractInvalid,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+      assertGrpcError(
+        errorBadPayload,
+        LedgerApiErrors.ConsistencyErrors.DisclosedContractInvalid,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
+    "EDMalformedDisclosedContracts",
+    "The ledger rejects malformed contract payloads",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
+
+      // This payload does not typecheck, it has different fields than the corresponding template
+      malformedArgument = Record(None, Seq(RecordField("", Some(Value(Value.Sum.Bool(false))))))
+
+      errorMalformedPayload <- testContext
+        .exerciseFetchDelegated(
+          testContext.disclosedContract
+            .update(_.arguments := malformedArgument)
+        )
+        .mustFail("using a malformed contract argument")
+
+      // Exercise a choice using an invalid disclosed contract (missing templateId)
+      errorMissingTemplateId <- testContext
+        .exerciseFetchDelegated(
+          testContext.disclosedContract
+            .update(_.modify(_.clearTemplateId))
+        )
+        .mustFail("using a disclosed contract with missing templateId")
+
+      // Exercise a choice using an invalid disclosed contract (missing contractId)
+      errorMissingContractId <- testContext
+        .exerciseFetchDelegated(
+          testContext.disclosedContract
+            .update(_.modify(_.clearTemplateId))
+        )
+        .mustFail("using a disclosed contract with missing contractId")
+    } yield {
+      assertGrpcError(
+        errorMalformedPayload,
+        // TODO ED: Verify that this error code is good enough for the user
```

Comment:
do we include the contract id in there?

Reply:
Good question. I'll defer addressing this to the actual implementation/mapping of error codes.